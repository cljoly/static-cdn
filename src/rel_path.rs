/* Copyright © 2024 Clément Joly
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

use std::path::Path;

use rusqlite::ToSql;

/// The database should hold paths relative to the folder walked through and the path part of urls
/// is also the relative path to that folder. Therefore, it makes sense to work most of the time
/// with those relative paths, as strings. The pair of types [`DbPath`] and [`DbPathBuilder`]
/// should make it easier to get that right, by keeping the root folder and returning relative
/// paths each time it's called. It also enforces that the right type is passed to DB functions.
#[derive(Debug)]
pub struct RelPath {
    // This is a relative path, to the same root folder, i.e. the base folder used for the walk
    rel_path: String,
}

impl RelPath {
    /// Returns a path relative to the root folder walked through
    pub fn get_relative_path(&self) -> &str {
        &self.rel_path
    }
}

impl ToSql for RelPath {
    fn to_sql(&self) -> rusqlite::Result<rusqlite::types::ToSqlOutput<'_>> {
        ToSql::to_sql(&self.rel_path)
    }
}

pub struct RelPathBuilder<'a> {
    root_folder: &'a Path,
}

impl<'a> RelPathBuilder<'a> {
    pub fn new<P>(root_folder: &'a P) -> Self
    where
        P: AsRef<Path> + ?Sized,
    {
        Self {
            root_folder: root_folder.as_ref(),
        }
    }

    pub fn db_path<P>(&'a self, child: &'a P) -> RelPath
    where
        P: AsRef<Path> + ?Sized,
    {
        let rel_path = child.as_ref().strip_prefix(self.root_folder).expect("directories are walked in from the root folder, they should be relative. This is a bug, please report");
        debug_assert!(
            rel_path.is_relative(),
            "{rel_path:?} should be relative for storage in DB"
        );

        RelPath {
            rel_path: format!("{rel_path:?}"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[should_panic]
    fn db_path_not_relative_to_root() {
        // This should not happen because we walk from the root dir, but still, test that this
        // panics
        let _ = RelPathBuilder::new("/made_up/for_testing")
            .db_path("./some_other_folder/some_other_file");
    }
}
