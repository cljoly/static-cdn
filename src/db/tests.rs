/* Copyright © 2024-2025 Clément Joly
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

use super::*;

use crate::rel_path::RelPathBuilder;

use anyhow::Result;

fn test_db_path() -> RelPath {
    RelPathBuilder::new("/made_up/for_testing")
        .db_path("/made_up/for_testing/some_other_folder/some_other_file")
}

#[test]
fn migrations() -> Result<()> {
    Ok(MIGRATIONS.validate()?)
}

#[test]
#[should_panic]
fn update_fails_when_nothing_exists() {
    let _ = open_transient().and_then(|mut c| {
        let _ = c.transaction().and_then(|tx| {
            // This should panic and nothing else can in this test
            let _ = update_metadata(&tx, &test_db_path(), &MetadataValues::default());
            Ok(())
        });
        Ok(())
    });
}

#[test]
fn insertion_and_checks() -> Result<()> {
    let db_path = test_db_path();
    let initial_metadata = MetadataValues {
        modified_since_epoch_sec: 12.,
        size: 10,
    };
    let updated_metadata = MetadataValues {
        size: 99,
        ..initial_metadata
    };
    let initial_checksum = Checksum::from(10);
    let updated_checksum = Checksum::from(20);
    let mut conn = open_transient()?;

    assert!(
        !exists_by_metadata(&mut conn, &db_path, &initial_metadata)?,
        "nothing should be inserted yet"
    );
    insta::assert_snapshot!("empty_table", read_all_files_rows(&conn));

    {
        let tx = conn.transaction()?;
        upsert_entry(&tx, &db_path, &initial_metadata, initial_checksum)?;
        tx.commit()?;
    }
    insta::assert_snapshot!("first_instert", read_all_files_rows(&conn));
    assert!(
        exists_by_metadata(&mut conn, &db_path, &initial_metadata)?,
        "should be inserted now"
    );
    assert!(
        exists_by_len_and_checksum(&mut conn, &db_path, &initial_metadata, initial_checksum)?,
        "should be inserted now, with the right checksum"
    );

    // Update
    {
        let tx = conn.transaction()?;
        upsert_entry(&tx, &db_path, &updated_metadata, updated_checksum)?;
        tx.commit()?;
    }
    insta::assert_snapshot!("after_update", read_all_files_rows(&conn));
    assert!(
        !exists_by_metadata(&mut conn, &db_path, &initial_metadata)?,
        "should not find the old version"
    );
    assert!(
        exists_by_metadata(&mut conn, &db_path, &updated_metadata)?,
        "should be updated"
    );
    assert!(
        exists_by_len_and_checksum(&mut conn, &db_path, &updated_metadata, updated_checksum)?,
        "should be updated, with the right checksum"
    );

    Ok(())
}
