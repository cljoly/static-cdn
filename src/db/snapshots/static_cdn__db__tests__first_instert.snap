---
source: src/db/tests.rs
expression: read_all_files_rows(&conn)
snapshot_kind: text
---
 SELECT * FROM files                                                                                                  
-------------------------------------------+--------------------------+-------------+---------------------------------
 path                                      | modified_since_epoch_sec | size        | checksum                        
 Text("some_other_folder/some_other_file") | Real(12.0)               | Integer(10) | Blob([10, 0, 0, 0, 0, 0, 0, 0])
